cornell-dti/course-plan
# AP/IB Overriden Requirements - Graph & Frontend Computation

### Summary 






This pull request is the second step towards implementing AP/IB overriding. [Notion](https://www.notion.so/courseplan/Discussion-Story-AP-IB-Infra-Update-to-handle-opt-out-and-requirement-overriding-ee2c3f234c6c44018531ed6ec34c56a1)  
It also sets up a lot of the architecture for handling general override functionality.

Regarding AP/IB equivalent courses:
- [x] added AP/IB equivalent course data to global data store (i.e., uniqueId mappings)
- [x] fixed repeated uniqueId bug for AP/IB equivalent courses
- [x] added dummy course for all exams taken that have no equivalent courses

Regarding the requirements algorithm/infrastructure:
- [x] added (AP/IB and general) user override choices to global data store
- [x] modified graph builder to add edges based on user's optIn override choices
- [x] implemented opting in/out of slot fulfillment during the frontend computation



Remaining TODOs:

- [ ] attach frontend modals/buttons
- [ ] add gtag analytics
- [ ] write additional unit tests for requirement graph builder and frontend util functions
- [ ] finish implementing general overriding next semester! :D



Depends on #489

### Test Plan 


1. Add an AP/IB exam normally.
1. Go to Firestore and add optIn/optOut options to the exam you want you opt in/out for. You may have to look through `requirements/data/` to find the right requirement names and slot names. For example:
![image](https://user-images.githubusercontent.com/34158284/118019709-4a810d00-b327-11eb-9d51-69fd059c1b03.png)
1. Check the requirements bar to ensure the opt in/out was successful.
1. Repeat for all the edge cases / combinations you can think of!

### Notes 


The schema for `AppOverridenRequirementChoices` is currently:
```
type AppOverridenRequirementChoices = Readonly<
  Record<
    number, // uniqueId
    {
      readonly optIn: Record<string, Set<string>>; // { 'requirementName': new Set(['slotName1', 'slotName2']) }
      readonly optOut: Record<string, Set<string>>; // { 'requirementName': new Set(['slotName1', 'slotName2']) }
    }
  >
>;
```
This may influence how we want to design the Firestore collection schema for general overrides.
